### ISSUE TITLE
Fix Generics for 8.0

### ISSUE THREAD

**Issue by treeowl** [haskell/containers#195]
I'm seeing

```
Data/Set/Base.hs:352:41: error:
    Not in scope: type constructor or class ‘NoSelector’
    Perhaps you meant ‘Selector’ (imported from GHC.Generics)
```

We need to fix this ASAP!

**Comment by kosmikus** [comment]
These new hand-written non-standard `Generic` instances seem to have been introduced comparatively recently. Is it clear that we really want them? I'm generally in favour of treating abstract types as abstract, and not having `Generic` instances for them.

The internal representation in GHC-8 has changed. Fixing them is relatively straight-forward, but will require CPP trickery to make work with both GHC-8 and older.

**Comment by treeowl** [comment]
They're non-standard to maintain the abstraction. Personally, I'm not really a big fan, but it's quite possible someone is relying on them, and I don't know that it's possible to define them better.

**Comment by kosmikus** [comment]
No, they clearly should be defined in the non-standard way they are, if they're to be defined at all. But I just looked at the original PR that introduced them which states

>    we want Generics for everything (otherwise we can't derive Generics
>    for any data types which contain these without creating orphan
>    instances)

but I don't think this justification is valid. We can very well derive `Generic` instances for datatypes containing abstract types, because `Generic` only defines a top-level transformation and does not touch the contained types.

I think it's unlikely that very many people rely on these (yet), because they don't seem to be in any released version of `containers`.

Anyway, it's not my decision to make. I can only say that I'd be leaning towards not having them at all.

This is what I think should be the diff for e.g. `Data.Set.Base`:
```
diff --git a/Data/Set/Base.hs b/Data/Set/Base.hs
index 47efe85..79be739 100644
--- a/Data/Set/Base.hs
+++ b/Data/Set/Base.hs
@@ -11,6 +11,9 @@
 {-# LANGUAGE TypeOperators #-}
 {-# LANGUAGE EmptyDataDecls #-}
 #endif
+#if __GLASGOW_HASKELL__ >= 800
+{-# LANGUAGE DataKinds #-}
+#endif
 
 #include "containers.h"
 
@@ -339,6 +342,8 @@ setDataType = mkDataType "Data.Set.Base.Set" [fromListConstr]
 {--------------------------------------------------------------------
   A Generic instance
 --------------------------------------------------------------------}
+#if __GLASGOW_HASKELL__ < 800
+
 data D1Set
 data C1Set
 
@@ -351,6 +356,24 @@ instance Constructor C1Set  where
 
 type Rep0Set a = D1 D1Set (C1 C1Set (S1 NoSelector (Rec0 [a])))
 
+#else
+
+type Rep0Set a =
+  D1 ('MetaData "Set" "Data.Set.Base" "containers" 'False) -- False = not a newtype
+    (C1 ('MetaCons "Set.fromList" 'PrefixI 'False) -- False = not a record
+      (S1
+        ('MetaSel
+          'Nothing -- Nothing = not a selector
+          'NoSourceUnpackedness
+          'NoSourceStrictness
+          'DecidedLazy
+        )
+        (Rec0 [a])
+      )
+    )
+
+#endif
+
 instance (Eq a, Ord a) => Generic (Set a) where
   type Rep (Set a) = Rep0Set a
   from s = M1 (M1 (M1 (K1 $ toList s)))
```

Introducing helper types like `D1Set` or `C1Set` is no longer required, neither are the instances for the `Datatype` and `Constructor` class. All meta-info is directly listed in the representation type, on the type level.

I unfortunately don't have time right now to fix this properly and do all the testing required. Perhaps someone else can apply the same to all the other modules and make a PR.

**Comment by treeowl** [comment]
I got confused by the `Data` instances, which *are* old. I'll take a look tomorrow at the history here, and probably consult the libraries list. Thanks for raising that concern.

**Comment by treeowl** [comment]
@kuk0, as @kosmikus indicates, you *don't* need these `Generic` instances to derive `Generic` instances for types containing containers! Since these potentially break efficiency intuition, and require careful tracking of `Generics` developments, I think we probably need good reasons to include them. Please send a message to libraries@haskell.org if you want these instances included, and explain why they are necessary. Otherwise, we may revert that change.

**Comment by cartazio** [comment]
I agree with @kosmikus,  as myself and the primary vector maintainers rejected a similar PR for Vector. High performance data structure do not have  *typical* or *effective* algebraic models (wrt any meaningful performance SLA), and any simulcra via the Generics rep will thus be ill behaved for some user. 

phrased differently: when providing a generics representation breaks some algorithmic contract the performance of the underlying data structure in a way that makes using generics no better than `fromList` and `toList`, thats an argument against supplying that instance.

**Comment by treeowl** [comment]
Fixed by removing the instances in #197
